openqa-bats-review

```diff
@@ -0,0 +1,207 @@
+    except RequestException as error:
+        log.error("%s: %s", url, error)
+        sys.exit(1)
+    return got.text
+
+
+@cache
+def get_job(url: str) -> dict:
+    """
+    Get a job from openQA
+    """
+    headers = {
+        "User-Agent": USER_AGENT,
+    }
+    try:
+        got = session.get(url, headers=headers, timeout=TIMEOUT)
+        got.raise_for_status()
+        data = got.json()
+    except RequestException as error:
+        log.error("%s: %s", url, error)
+        sys.exit(1)
+    return data["job"]
+
+
+def grep_notok(url: str) -> set[str]:
+    """
+    Grep for "not ok" lines and return a set with the failing tests
+    prefixed by the filename
+    """
+    notok = set()
+    # Note: We use a prefix because some tests have more than 1 TAP file
+    prefix = os.path.basename(url)
+    data = get_file(url)
+    lines = data.splitlines()
+    # Fetch the plan to check that we don't have a truncated TAP file
+    last = 0
+    for line in lines:
+        if line.startswith("1.."):
+            last = int(line.split("..", 1)[1])
+            break
+    if not last:
+        log.error("Malformed TAP file: %s", url)
+        sys.exit(1)
+    # Trust the plan
+    tests = 0
+    for line in lines:
+        if not line.startswith(("ok", "not ok", "#not ok")):
+            continue
+        tests += 1
+        try:
+            test = NOT_OK.findall(line)[0]
+            notok.add(f"{prefix}:{test}")
+        except IndexError:
+            continue
+    if tests != last:
+        log.error("Truncated TAP file: %s", url)
+        sys.exit(1)
+    return notok
+
+
+def process_files(files: list[str]) -> set[str]:
+    """
+    Process TAP files
+    """
+    # The test for these packages have only one TAP file: aardvark-dns & netavark
+    if len(files) == 1:
+        return grep_notok(files[0])
+    # Use multithreading for tests like buildah, runc & skopeo (2 files) & podman (4 files)
+    with ThreadPoolExecutor(max_workers=len(files)) as executor:
+        return set(itertools.chain.from_iterable(executor.map(grep_notok, files)))
+
+
+def get_clone_chain(openqa_host: str, job_id: int) -> list[int]:
+    """
+    Follow clones recursively and return the full chain:
+    [job_id, origin_id, origin_id_of_origin, ...]
+    """
+    chain = []
+    current: int | None = job_id
+    while current:
+        # We use "/details" because we'll need this information again and get_job() is cached
+        job = get_job(f"{openqa_host}/api/v1/jobs/{current}/details")
+        if "BATS_PACKAGE" not in job["settings"]:
+            log.error("Not a BATS test: %d", job_id)
+            sys.exit(1)
+        chain.append(current)
+        current = job.get("origin_id")
+    return chain
+
+
+def main(url: str) -> None:
+    """
+    Main function
+    """
+    if not url.startswith(("http://", "https://")):
+        url = f"https://{url}"
+    urlx = urlparse(url)
+    openqa_host = f"{urlx.scheme}://{urlx.netloc}"
```

Comment:
```suggestion
    urlx = urlparse(url)
    if all([u.scheme,u.netloc]):
        openqa_host=f"{urlx.scheme}://{urlx.netloc}"
    else:
        openqa_host=f"https://{url}"
```
Either way I am not blocking you on this.

Reply:
But this needs 5 lines instead of 4 and I find my version more readable (thouth maybe because I use this pattern everywhere).

Ideally, passing `scheme` to `urlparse()` should work but we get this instead:

```
>>> urlparse("openqa.suse.de/abc", scheme="https")
ParseResult(scheme='https', netloc='', path='openqa.suse.de/abc', params='', query='', fragment='')
```

Reply:
Sure. no probs